File: python/src/main/java/org/digma/intellij/plugin/psi/python/PythonLanguageService.java
package org.digma.intellij.plugin.psi.python;

public class PythonLanguageService{

//    private static final Logger LOGGER = Logger.getInstance(PythonLanguageService.class);
//
//    private final Project project;
//
//    private final ProjectFileIndex projectFileIndex;

    /*
    It's better, as much as possible, in language services especially, not to initialize service dependencies in the constructor but use
    a getInstance for services when they are first needed. that will minimize the possibility for cyclic dependencies.
     */
//    public PythonLanguageService(Project project) {
//        this.project = project;
//        projectFileIndex = project.getService(ProjectFileIndex.class);
//    }
//
//
//    @Override
//    public void ensureStartupOnEDT(@NotNull Project project) {
//        //nothing to do
//    }
//
//    @Override
//    public void runWhenSmart(Runnable task) {
//        if (DumbService.isDumb(project)) {
//            DumbService.getInstance(project).runWhenSmart(task);
//        } else {
//            task.run();
//        }
//    }
//
//    @Nullable
//    @Override
//    public Language getLanguageForMethodCodeObjectId(@NotNull String methodId) {
//
//        var functionName = PythonLanguageUtils.extractFunctionNameFromCodeObjectId(methodId);
//
//        //try to find a function that produces the same code object id,
//        // if found return its language, else null
//        return ReadAction.compute(() -> {
//            var functions = PyFunctionNameIndex.find(functionName, project, GlobalSearchScope.projectScope(project));
//
//            for (PyFunction function : functions) {
//                var codeObjectId = PythonLanguageUtils.createPythonMethodCodeObjectId(project, function);
//                //python method has multiple ids, we don't know what the backend will send so check with all possible ids
//                List<String> allIds = PythonAdditionalIdsProvider.getAdditionalIdsInclusive(codeObjectId, false);
//                if (allIds.contains(methodId)) {
//                    return function.getLanguage();
//                }
//            }
//
//            return null;
//        });
//    }
//
//
//    @Override
//    public @Nullable Language getLanguageForClass(@NotNull String className) {
//        return ReadAction.compute(() -> {
//            var classes = PyClassNameIndex.find(className, project, GlobalSearchScope.projectScope(project));
//            var cls = classes.stream().findFirst();
//            return cls.map(PsiElement::getLanguage).orElse(null);
//        });
//    }
//
//    @Override
//    public boolean isSupportedFile(@NotNull Project project, @NotNull VirtualFile newFile) {
//        if (!VfsUtilsKt.isValidVirtualFile(newFile)) {
//            return false;
//        }
//        PsiFile psiFile = PsiAccessUtilsKt.runInReadAccessWithResult(() -> PsiManager.getInstance(project).findFile(newFile));
//        return PsiUtils.isValidPsiFile(psiFile) && isSupportedFile(psiFile);
//    }
//
//    @Override
//    public boolean isSupportedFile(@NotNull PsiFile psiFile) {
//        return PythonLanguage.INSTANCE.equals(psiFile.getLanguage());
//    }
//
//    @Override
//    @NotNull
//    public MethodUnderCaret detectMethodUnderCaret(@NotNull Project project, @NotNull PsiFile psiFile, Editor selectedEditor, int caretOffset) {
//
//        //detectMethodUnderCaret should run very fast and return a result,
//        // this operation may become invalid very soon if user clicks somewhere else.
//        // no retry because it needs to complete very fast
//        //it may be called from EDT or background, runInReadAccessWithResult will acquire read access if necessary.
//        return executeCatchingWithResult(() -> PsiAccessUtilsKt.runInReadAccessWithResult(() -> detectMethodUnderCaret(project, psiFile, caretOffset)), throwable -> {
//            ErrorReporter.getInstance().reportError(project, getClass().getSimpleName() + ".detectMethodUnderCaret", throwable);
//            return new MethodUnderCaret("", "", "", "", PsiUtils.psiFileToUri(psiFile), caretOffset, null, false);
//        });
//    }
//
//    private MethodUnderCaret detectMethodUnderCaret(@NotNull Project project, @NotNull PsiFile psiFile, int caretOffset) {
//
//        if (!isSupportedFile(psiFile)) {
//            return new MethodUnderCaret("", "", "", "", PsiUtils.psiFileToUri(psiFile), caretOffset, null, false);
//        }
//        PsiElement underCaret = psiFile.findElementAt(caretOffset);
//        if (underCaret == null) {
//            return new MethodUnderCaret("", "", "", "", PsiUtils.psiFileToUri(psiFile), caretOffset, null, true);
//        }
//        PyFunction pyFunction = PsiTreeUtil.getParentOfType(underCaret, PyFunction.class);
//        if (pyFunction != null) {
//            var methodId = PythonLanguageUtils.createPythonMethodCodeObjectId(project, pyFunction);
//            var name = pyFunction.getName() == null ? "" : pyFunction.getName();
//            var containingClass = pyFunction.getContainingClass();
//            var className = containingClass == null ? "" : containingClass.getName() + ".";
//            return new MethodUnderCaret(methodId, name, className, "", PsiUtils.psiFileToUri(psiFile), caretOffset);
//        }
//        return new MethodUnderCaret("", "", "", "", PsiUtils.psiFileToUri(psiFile), caretOffset);
//    }
//
//    /**
//     * navigate to a method. this method is meant to be used only to navigate to a method in the current selected editor.
//     * it is used from the methods preview list. it will not navigate to any method in the project.
//     */
//    @Override
//    public void navigateToMethod(String methodId) {
//
//        Log.log(LOGGER::debug, "got navigate to method request {}", methodId);
//
//        var functionName = PythonLanguageUtils.extractFunctionNameFromCodeObjectId(methodId);
//
//
//        ReadActions.ensureReadAction(() -> {
//            var functions = PyFunctionNameIndex.find(functionName, project, GlobalSearchScope.allScope(project));
//
//            //PyFunctionNameIndex may find many functions, we want only one, so if we found it break the loop
//            for (PyFunction function : functions) {
//                if (function.isValid()) {
//                    var codeObjectId = PythonLanguageUtils.createPythonMethodCodeObjectId(project, function);
//                    List<String> allIds = PythonAdditionalIdsProvider.getAdditionalIdsInclusive(codeObjectId, false);
//                    if (allIds.contains(methodId) && (function.canNavigateToSource())) {
//                        Log.log(LOGGER::debug, "navigating to method {}", function);
//                        function.navigate(true);
//                        break;
//                    }
//                }
//            }
//        });
//
//    }
//
//    @Override
//    public boolean isServiceFor(@NotNull Language language) {
//        return PythonLanguage.class.equals(language.getClass());
//    }
//
//    @Override
//    public Map<String, String> findWorkspaceUrisForCodeObjectIdsForErrorStackTrace(List<String> codeObjectIds) {
//
//        var workspaceUris = new HashMap<String, String>();
//
//        codeObjectIds.forEach(methodId -> {
//
//            var functionName = PythonLanguageUtils.extractFunctionNameFromCodeObjectId(methodId);
//
//            ReadActions.ensureReadAction(() -> {
//                var functions = PyFunctionNameIndex.find(functionName, project, GlobalSearchScope.projectScope(project));
//
//                //PyFunctionNameIndex may find many functions, we want only one, so if we found it break the loop
//                // assuming that our createPythonMethodCodeObjectId returns a unique id
//                for (PyFunction function : functions) {
//                    if (function.isValid()) {
//                        PsiFile psiFile = function.getContainingFile();
//                        if (PythonLanguageUtils.isProjectFile(project, psiFile)) {
//                            var codeObjectId = PythonLanguageUtils.createPythonMethodCodeObjectId(project, function);
//                            List<String> allIds = PythonAdditionalIdsProvider.getAdditionalIdsInclusive(codeObjectId, false);
//                            if (allIds.contains(methodId)) {
//                                String url = PsiUtils.psiFileToUri(psiFile);
//                                workspaceUris.put(codeObjectId, url);
//                                break;
//                            }
//                        }
//                    }
//                }
//            });
//        });
//
//        return workspaceUris;
//    }
//
//    @Override
//    public Map<String, Pair<String, Integer>> findWorkspaceUrisForMethodCodeObjectIds(List<String> methodCodeObjectIds) {
//
//        Map<String, Pair<String, Integer>> workspaceUris = new HashMap<>();
//
//        methodCodeObjectIds.forEach(methodId -> {
//
//            var functionName = PythonLanguageUtils.extractFunctionNameFromCodeObjectId(methodId);
//
//            ReadActions.ensureReadAction(() -> {
//                var functions = PyFunctionNameIndex.find(functionName, project, GlobalSearchScope.projectScope(project));
//
//                //PyFunctionNameIndex may find many functions, we want only one, so if we found it break the loop.
//                // assuming that our createPythonMethodCodeObjectId returns a unique id
//                for (PyFunction function : functions) {
//                    if (function.isValid()) {
//                        PsiFile psiFile = function.getContainingFile();
//                        if (PythonLanguageUtils.isProjectFile(project, psiFile)) {
//                            var codeObjectId = PythonLanguageUtils.createPythonMethodCodeObjectId(project, function);
//                            List<String> allIds = PythonAdditionalIdsProvider.getAdditionalIdsInclusive(codeObjectId, false);
//                            if (allIds.contains(methodId)) {
//                                String url = PsiUtils.psiFileToUri(psiFile);
//                                workspaceUris.put(methodId, new Pair<>(url, function.getTextOffset()));
//                                break;
//                            }
//                        }
//                    }
//                }
//            });
//        });
//
//        return workspaceUris;
//    }
//
//    @Override
//    public Map<String, Pair<String, Integer>> findWorkspaceUrisForSpanIds(List<String> spanIds) {
//        return PythonSpanNavigationProvider.getInstance(project).getUrisForSpanIds(spanIds);
//    }
//
//    @Override
//    public Set<EndpointInfo> lookForDiscoveredEndpoints(String endpointId) {
//        return Collections.emptySet();
//    }
//
//    @Override
//    public @NotNull DocumentInfo buildDocumentInfo(@NotNull PsiFileCachedValueWithUri psiFileCachedValue, BuildDocumentInfoProcessContext context) {
//
//        var psiFile = psiFileCachedValue.getValue();
//        if (!PsiUtils.isValidPsiFile(psiFile)) {
//            return new DocumentInfo(psiFileCachedValue.getUri(), new HashMap<>());
//        }
//
//        //todo: when python support is revived pass psiFileCachedValue to PythonCodeObjectsDiscovery.buildDocumentInfo
//        // and use PsiFileCachedValueWithUri to always use a valid file
//
//        Log.log(LOGGER::debug, "got buildDocumentInfo request for {}", psiFile);
//        if (psiFile instanceof PyFile pyFile) {
//
//            return ProgressManager.getInstance().runProcess(() -> Retries.retryWithResult(() ->
//                            ReadAction.compute(() -> PythonCodeObjectsDiscovery.buildDocumentInfo(project, pyFile)),
//                    Throwable.class, 50, 5), new EmptyProgressIndicator());
//
//        } else {
//            Log.log(LOGGER::debug, "psi file is noy python, returning empty DocumentInfo for {}", psiFile);
//            return new DocumentInfo(PsiUtils.psiFileToUri(psiFile), new HashMap<>());
//        }
//    }
//
//    @Override
//    public @NotNull DocumentInfo buildDocumentInfo(@NotNull PsiFileCachedValueWithUri psiFileCachedValue, @Nullable FileEditor newEditor, BuildDocumentInfoProcessContext context) {
//        return buildDocumentInfo(psiFileCachedValue, context);
//    }
//
//
//    @Override
//    public boolean isRelevant(VirtualFile file) {
//
//        return PsiAccessUtilsKt.runInReadAccessWithResult(() -> {
//            if (file.isDirectory() || !VfsUtilsKt.isValidVirtualFile(file)) {
//                return false;
//            }
//
//            PsiFile psiFile = PsiManager.getInstance(project).findFile(file);
//            if (PsiUtils.isValidPsiFile(psiFile)) {
//                return false;
//            }
//
//            return isRelevant(psiFile);
//        });
//    }
//
//
//    @Override
//    public boolean isRelevant(PsiFile psiFile) {
//
//        return SlowOperationsUtilsKt.allowSlowOperation(() -> PsiAccessUtilsKt.runInReadAccessWithResult(() -> psiFile.isValid() &&
//                psiFile.isWritable() &&
//                PythonLanguageUtils.isProjectFile(project, psiFile) &&
//                !projectFileIndex.isInLibrary(psiFile.getVirtualFile()) &&
//                !projectFileIndex.isExcluded(psiFile.getVirtualFile()) &&
//                isSupportedFile(psiFile)));
//    }
//
//    @Override
//    public void refreshMethodUnderCaret(@NotNull Project project, @NotNull PsiFile psiFile, @Nullable Editor selectedEditor, int offset) {
//        MethodUnderCaret methodUnderCaret = detectMethodUnderCaret(project, psiFile, selectedEditor, offset);
//        CaretContextService.getInstance(project).contextChanged(methodUnderCaret);
//    }
//
//    @Override
//    public boolean isCodeVisionSupported() {
//        return true;
//    }
//
//    @Override
//    public @Nullable PsiElement getPsiElementForMethod(@NotNull String methodId) {
//        var functionName = PythonLanguageUtils.extractFunctionNameFromCodeObjectId(methodId);
//
//        //try to find a function that produces the same code object id,
//        // if found return its language, else null
//        return ReadAction.compute(() -> {
//            var functions = PyFunctionNameIndex.find(functionName, project, GlobalSearchScope.projectScope(project));
//
//            for (PyFunction function : functions) {
//                var codeObjectId = PythonLanguageUtils.createPythonMethodCodeObjectId(project, function);
//                //python method has multiple ids, we don't know what the backend will send so check with all possible ids
//                List<String> allIds = PythonAdditionalIdsProvider.getAdditionalIdsInclusive(codeObjectId, false);
//                if (allIds.contains(methodId)) {
//                    return function;
//                }
//            }
//
//            return null;
//        });
//    }
//
//    @Override
//    public @Nullable PsiElement getPsiElementForClassByMethodId(@NotNull String methodId) {
//        var functionName = PythonLanguageUtils.extractFunctionNameFromCodeObjectId(methodId);
//
//        //try to find a function that produces the same code object id,
//        // if found return its language, else null
//        return ReadAction.compute(() -> {
//            var functions = PyFunctionNameIndex.find(functionName, project, GlobalSearchScope.projectScope(project));
//
//            for (PyFunction function : functions) {
//                var codeObjectId = PythonLanguageUtils.createPythonMethodCodeObjectId(project, function);
//                //python method has multiple ids, we don't know what the backend will send so check with all possible ids
//                List<String> allIds = PythonAdditionalIdsProvider.getAdditionalIdsInclusive(codeObjectId, false);
//                if (allIds.contains(methodId)) {
//                    return PsiTreeUtil.getParentOfType(function, PyClass.class);
//                }
//            }
//
//            return null;
//        });
//    }
//
//    @Override
//    public @Nullable PsiElement getPsiElementForClassByName(@NotNull String className) {
//        return ReadAction.compute(() -> {
//            var classes = PyClassNameIndex.find(className, project, GlobalSearchScope.projectScope(project));
//            var cls = classes.stream().findFirst();
//            return cls.orElse(null);
//        });
//    }
//
//    @Override
//    public @NotNull InstrumentationProvider getInstrumentationProvider() {
//        return new NoOpInstrumentationProvider();
//    }
//
//
//    //this method is called only from CodeLensService, CodeLensService should handle exceptions
//    @Override
//    public @NotNull Map<String, PsiElement> findMethodsByCodeObjectIds(@NotNull PsiFile psiFile, @NotNull List<String> methodIds) {
//        if (methodIds.isEmpty() || !PsiUtils.isValidPsiFile(psiFile)) {
//            return Collections.emptyMap();
//        }
//
//        return PsiAccessUtilsKt.runInReadAccessWithResult((Computable<Map<String, PsiElement>>) () -> {
//            var methods = new HashMap<String, PsiElement>();
//            var traverser = SyntaxTraverser.psiTraverser(psiFile);
//            for (PsiElement element : traverser) {
//                if (element instanceof PyFunction pyFunction) {
//                    var codeObjectId = PythonLanguageUtils.createPythonMethodCodeObjectId(psiFile.getProject(), pyFunction);
//                    var additionalIds = PythonAdditionalIdsProvider.getAdditionalIdsInclusive(codeObjectId, false);
//                    if (methodIds.stream().anyMatch(additionalIds::contains)) {
//                        methods.put(codeObjectId, element);
//                    }
//                }
//            }
//
//            return methods;
//        });
//    }
}

File: python/src/main/kotlin/org/digma/intellij/plugin/python/PythonLanguageService.kt
package org.digma.intellij.plugin.python

import com.intellij.lang.Language
import com.intellij.openapi.application.readAction
import com.intellij.openapi.application.smartReadAction
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.diagnostic.thisLogger
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.fileTypes.FileType
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.PsiManager
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.util.concurrency.annotations.RequiresReadLock
import com.jetbrains.python.PythonFileType
import com.jetbrains.python.PythonLanguage
import com.jetbrains.python.psi.PyFile
import com.jetbrains.python.psi.PyFunction
import com.jetbrains.python.psi.stubs.PyClassNameIndex
import com.jetbrains.python.psi.stubs.PyFunctionNameIndex
import kotlinx.coroutines.ensureActive
import org.digma.intellij.plugin.common.DisposableAdaptor
import org.digma.intellij.plugin.common.EDT
import org.digma.intellij.plugin.common.ReadActions
import org.digma.intellij.plugin.common.isProjectValid
import org.digma.intellij.plugin.common.isValidVirtualFile
import org.digma.intellij.plugin.common.suspendableRetry
import org.digma.intellij.plugin.discovery.FileDiscoveryProvider
import org.digma.intellij.plugin.discovery.model.EndpointLocation
import org.digma.intellij.plugin.instrumentation.InstrumentationProvider
import org.digma.intellij.plugin.instrumentation.NoOpInstrumentationProvider
import org.digma.intellij.plugin.log.Log
import org.digma.intellij.plugin.model.discovery.DocumentInfo
import org.digma.intellij.plugin.model.discovery.MethodUnderCaret
import org.digma.intellij.plugin.psi.LanguageService
import org.digma.intellij.plugin.psi.PsiUtils
import org.digma.intellij.plugin.python.discovery.PythonFileDiscoveryProvider
import org.digma.intellij.plugin.python.discovery.findFunctionByMethodId
import org.digma.intellij.plugin.python.navigation.PythonSpanNavigationProvider
import java.util.function.Supplier
import kotlin.coroutines.coroutineContext

@Suppress("LightServiceMigrationCode", "UnstableApiUsage")
class PythonLanguageService(private val project: Project) : LanguageService, DisposableAdaptor {

    private val logger: Logger = thisLogger()

    override fun getLanguage(): Language {
        return PythonLanguage.INSTANCE
    }

    override fun getFileType(): FileType {
        return PythonFileType.INSTANCE
    }

    override fun isSupportedFile(psiFile: PsiFile): Boolean {
        return ReadActions.ensureReadAction(Supplier {
            psiFile is PyFile &&
                    PsiUtils.isValidPsiFile(psiFile) &&
                    PythonLanguage.INSTANCE == psiFile.viewProvider.baseLanguage
        })
    }

    override suspend fun detectMethodUnderCaret(
        virtualFile: VirtualFile,
        editor: Editor,
        caretOffset: Int
    ): MethodUnderCaret {

        Log.trace(logger, project, "detectMethodUnderCaret called {}", virtualFile)

        if (!isSupportedFile(virtualFile)) {
            return MethodUnderCaret.empty(virtualFile.url)
        }

        if (!isValidVirtualFile(virtualFile)) {
            return MethodUnderCaret.EMPTY
        }

        return smartReadAction(project) {
            return@smartReadAction detectMethodUnderCaretImpl(virtualFile, caretOffset)
        }
    }

    @RequiresReadLock(generateAssertion = false)
    private fun detectMethodUnderCaretImpl(
        virtualFile: VirtualFile,
        caretOffset: Int
    ): MethodUnderCaret {

        val psiFile = PsiManager.getInstance(project).findFile(virtualFile)?.takeIf { it.isValid && virtualFile.isValid }
            ?: return MethodUnderCaret.empty(virtualFile.url)

        val underCaret: PsiElement = psiFile.findElementAt(caretOffset).takeIf { psiFile.isValid && virtualFile.isValid }
            ?: return MethodUnderCaret.empty(virtualFile.url)

        val pyFunction = underCaret.takeIf { psiFile.isValid && virtualFile.isValid }?.let {
            PsiTreeUtil.getParentOfType(underCaret, PyFunction::class.java)
        } ?: return MethodUnderCaret.empty(virtualFile.url)

        val methodId = createPythonMethodCodeObjectId(project, pyFunction)
        val name = pyFunction.name ?: ""
        val containingClassName = pyFunction.containingClass?.name ?: ""
        return MethodUnderCaret(methodId, name, containingClassName, "", virtualFile.url, caretOffset)

    }

    override suspend fun buildDocumentInfo(virtualFile: VirtualFile): DocumentInfo? {

        Log.trace(logger, project, "buildDocumentInfo called {}", virtualFile)

        //don't call this method on EDT or in read access, read access is acquired when needed
        EDT.assertNonDispatchThread()
        ReadActions.assertNotInReadAccess()

        Log.trace(logger, project, "got buildDocumentInfo request for {}", virtualFile)

        if (!isValidVirtualFile(virtualFile)) {
            Log.trace(logger, project, "buildDocumentInfo: virtualFile is not valid for {}", virtualFile)
            return null
        }
        val psiFile = smartReadAction(project) {
            val psiFile = PsiManager.getInstance(project).findFile(virtualFile)
            if (psiFile == null) {
                Log.trace(logger, project, "buildDocumentInfo: could not find psiFile for {}", virtualFile)
                null
            } else if (psiFile !is PyFile) {
                Log.trace(logger, project, "buildDocumentInfo: psiFile is not PyFile {}", virtualFile)
                null
            } else if (!PsiUtils.isValidPsiFile(psiFile)) {
                Log.trace(logger, project, "buildDocumentInfo: psiFile is not valid for {}", virtualFile)
                null
            } else if (!isSupportedFile(psiFile)) {
                Log.trace(logger, project, "buildDocumentInfo: psiFile is not supported for {}", virtualFile)
                null
            } else if (!isProjectValid(project)) {
                Log.trace(logger, project, "buildDocumentInfo: project is not valid for {}", virtualFile)
                null
            } else {
                psiFile
            }
        }
        coroutineContext.ensureActive()
        return psiFile?.let {
            suspendableRetry {
                coroutineContext.ensureActive()
                org.digma.intellij.plugin.python.discovery.buildDocumentInfo(project, it, virtualFile.url, getLanguage())
            }
        }
    }

    override suspend fun getLanguageForMethodCodeObjectId(methodId: String): Language? {
        val methodName = methodId.substringAfter("\$_$")

        val function = readAction {
            PyFunctionNameIndex.find(methodName, project, GlobalSearchScope.projectScope(project))
                .firstOrNull { createPythonMethodCodeObjectId(project, it) == methodId }
        }

        return function?.let {
            PythonLanguage.INSTANCE
        }
    }

    override suspend fun getLanguageForClass(className: String): Language? {
        val classNameToFind = className.substringAfterLast(".")
        val clazz = readAction {
            PyClassNameIndex.find(classNameToFind, project, GlobalSearchScope.projectScope(project))
                .firstOrNull { it.qualifiedName == className }
        }

        return clazz?.let {
            PythonLanguage.INSTANCE
        }
    }

    override suspend fun detectMethodBySpan(project: Project, spanCodeObjectId: String): String? {
        return PythonSpanNavigationProvider.getInstance(project).getMethodIdBySpanId(spanCodeObjectId)
    }

    override suspend fun navigateToMethod(methodId: String) {
        val methodName = getPythonMethodNameFromMethodId(methodId)
        val function = readAction {
            PyFunctionNameIndex.find(methodName, project, GlobalSearchScope.projectScope(project))
                .firstOrNull { createPythonMethodCodeObjectId(project, it) == methodId }
        }

        function?.takeIf { it.canNavigate() }.let {
            it?.navigate(true)
        }
    }

    override suspend fun findWorkspaceUrisForCodeObjectIdsForErrorStackTrace(methodCodeObjectIds: List<String>): Map<String, String> {

        val workspaceUrls = mutableMapOf<String, String>()

        methodCodeObjectIds.filter { isPythonMethodId(it) }.forEach { methodId ->
            val url = smartReadAction(project) {
                val function = findFunctionByMethodId(project, methodId) { GlobalSearchScope.projectScope(project) }
                function?.containingFile?.virtualFile?.url
            }
            url?.let {
                workspaceUrls[methodId] = it
            }
        }

        return workspaceUrls
    }

    override suspend fun findWorkspaceUrisForMethodCodeObjectIds(methodCodeObjectIds: List<String>): Map<String, Pair<String, Int>> {

        val workspaceUrls = mutableMapOf<String, Pair<String, Int>>()

        methodCodeObjectIds.filter { isPythonMethodId(it) }.forEach { methodId ->
            val pair = smartReadAction(project) {
                val function = findFunctionByMethodId(project, methodId) { GlobalSearchScope.projectScope(project) }
                function?.let { Pair(it.containingFile.virtualFile.url, it.textOffset) }
            }
            pair?.let {
                workspaceUrls[methodId] = it
            }
        }

        return workspaceUrls
    }

    override suspend fun findWorkspaceUrisForSpanIds(spanIds: List<String>): Map<String, Pair<String, Int>> {
        return PythonSpanNavigationProvider.getInstance(project).getUrisForSpanIds(spanIds)
    }

    override fun lookForDiscoveredEndpoints(endpointId: String): Set<EndpointLocation> {
        return setOf()
    }

    override fun isCodeVisionSupported(): Boolean = true

    override suspend fun findMethodPsiElementByMethodId(methodId: String): PsiElement? {
        return smartReadAction(project) {
            findFunctionByMethodId(project, methodId) { GlobalSearchScope.projectScope(project) }
        }
    }

    override suspend fun findClassPsiElementByMethodId(methodId: String): PsiElement? {
        return smartReadAction(project) {
            val function = findFunctionByMethodId(project, methodId) { GlobalSearchScope.projectScope(project) }
            function?.containingClass
        }
    }

    override suspend fun findClassPsiElementByClassName(className: String): PsiElement? {
        return smartReadAction(project) {
            PyClassNameIndex.findByQualifiedName(className, project, GlobalSearchScope.projectScope(project))
                .firstOrNull { it.isValid }
        }
    }

    override fun getInstrumentationProvider(): InstrumentationProvider {
        return NoOpInstrumentationProvider()
    }

    override fun findMethodsByCodeObjectIds(
        psiFile: PsiFile,
        methodIds: List<String>
    ): Map<String, PsiElement> {

        if (methodIds.isEmpty() || !PsiUtils.isValidPsiFile(psiFile)) {
            return emptyMap()
        }

        val functionNames = methodIds.associateBy { it.substringAfter("\$_$") }

        return ReadActions.ensureReadAction(Supplier {
            PsiTreeUtil.collectElementsOfType(psiFile, PyFunction::class.java)
                .filter { it.name != null && it.name in functionNames.keys }
                .associateBy { functionNames[it.name as String] as String }
        }) ?: emptyMap()

    }

    override fun getDiscoveryProvider(): FileDiscoveryProvider {
        return PythonFileDiscoveryProvider()
    }
}